src/simple_tracker_track_provider/simple_tracker_track_provider/track_provider_node.py

```diff
@@ -12,88 +12,108 @@
-      track_array_msg.tracks = [self._track_to_msg(tracker) for tracker in self.video_tracker.live_trackers]
-      track_array_msg.frame = self.msg_frame.frame
-      self.pub_tracker_tracks.publish(track_array_msg)
-      
       tracking_msg = TrackingState()
-      tracking_msg.epoch = self.msg_frame.epoch
-      tracking_msg.fps = self.msg_frame.fps
-      tracking_msg.frame_count = self.msg_frame.frame_count
+      tracking_msg.header = msg_frame.header
       tracking_msg.trackable = sum(map(lambda x: x.is_tracking(), self.video_tracker.live_trackers))
       tracking_msg.alive = len(self.video_tracker.live_trackers)
       tracking_msg.started = self.video_tracker.total_trackers_started
       tracking_msg.ended = self.video_tracker.total_trackers_finished
-      self.pub_tracker_tracking_state.publish(tracking_msg)
 
+      self.pub_tracker_detects.publish(detect_array_msg)
+      self.pub_tracker_trajectories.publish(trajectory_array_msg)
+      self.pub_tracker_tracking_state.publish(tracking_msg)
 
-  def _msg_to_bbox(self, bbox_msg):
-    return bbox_msg.x, bbox_msg.y, bbox_msg.w, bbox_msg.h
+  def _msg_to_bbox(self, bbox_msg: BoundingBox2D):
+    x, y, w, h = (int(bbox_msg.center.position.x - (bbox_msg.size_x / 2)), int(bbox_msg.center.position.y - (bbox_msg.size_y / 2)), int(bbox_msg.size_x), int(bbox_msg.size_y))
+    return x, y, w, h
 
-  def _track_to_msg(self, tracker):
+  def _detects_to_msg(self, tracker):
 
     x, y, w, h = tracker.get_bbox()
-    bbox_msg = BoundingBox()
-    bbox_msg.x = x
-    bbox_msg.y = y
-    bbox_msg.w = w
-    bbox_msg.h = h
+    bbox_msg = BoundingBox2D()
+    bbox_msg.center.position.x = float(x+(w/2))
+    bbox_msg.center.position.y = float(y+(h/2))
+    bbox_msg.center.theta = 0.0
+    bbox_msg.size_x = float(w)
+    bbox_msg.size_y = float(h)
+
+    detect_msg = Detection2D()
+    detect_msg.id = f'{tracker.id}-{tracker.tracking_state}'
+    detect_msg.bbox = bbox_msg
+
+    return detect_msg
+
+  def _trajectories_to_msg(self, tracker):
 
-    track_msg = Track()
-    track_msg.id = tracker.id
-    track_msg.state = tracker.tracking_state
-    track_msg.bbox = bbox_msg
+    track_msg = TrackTrajectory()
+    track_msg.id = f'{tracker.id}-{tracker.tracking_state}'
 
     for center_point in tracker.center_points:
       (x,y) = center_point[0]
-      state = center_point[1]
-      cpoint = CenterPoint()
-      cpoint.x = x
-      cpoint.y = y
-      cpoint.state = state
-      track_msg.path.append(cpoint)
+      state = center_point[1]      
+      point = TrackPoint()
+      point.center.x = float(x)
+      point.center.y = float(y)
+      point.state = state
+      point.prediction = False
+      track_msg.trajectory.append(point)
+
+    for center_point in tracker.predictor_center_points:
+      (x,y) = center_point
+      point = TrackPoint()
+      point.center.x = float(x)
+      point.center.y = float(y)
+      point.prediction = True
+      track_msg.trajectory.append(point)
```

Comment:
These predicted points don't actually correspond to tracked frames right?

Why do we combine them in a `TrackTrajectory`?

Reply:
Yeah, so again, this is an approach to try and standardise message types, but agree this needs to be refactored as I am not happy with it at the moment. It works, but I think the cleaner approach would be including a new topic that publishes predictions, you can then decide if you want to subscribe to predictions etc versus getting them as part of the existing trajectory which will be very confusing. So yes, agreed, this needs to change.

Reply:
I have now refactored this in order to publish the predictions to its own end point/action. I am happier with this, however when rendering the annotated frame it now does require an additional for loop.